app/src/main/java/com/willowtree/vocable/presets/RoomPresetCategoriesRepository.kt

```diff
@@ -1,57 +1,82 @@
 import kotlinx.coroutines.sync.Mutex
 import kotlinx.coroutines.sync.withLock
 
 class RoomPresetCategoriesRepository(
-    context: Context
+    private val database: VocableDatabase
 ) : PresetCategoriesRepository {
     // Needed because we're checking database contents and making updates based
     // on them. Multiple callers may call at once.
     private val categoryMutex = Mutex()
-    private val database = VocableDatabase.getVocableDatabase(context)
 
-    override suspend fun getPresetCategories(): List<Category.PresetCategory> {
-        return categoryMutex.withLock {
-            val presetDtos = database.presetCategoryDao().getPresetCategoryDtos()
+    override fun getPresetCategories(): Flow<List<Category.PresetCategory>> {
+        return database.presetCategoryDao().getAllPresetCategoriesFlow().map {
+            it.map {
+                Category.PresetCategory(
+                    it.categoryId,
+                    it.sortOrder,
+                    it.hidden,
+                    PresetCategories.values()
+                        .first { presetCategory -> presetCategory.id == it.categoryId }
+                        .getNameId()
+                )
+            }
+        }.onStart { ensurePopulated() }
+    }
+
+    private suspend fun ensurePopulated() {
+        categoryMutex.withLock {
+            val dbPresets = database.presetCategoryDao().getAllPresetCategoriesFlow().first()
             PresetCategories.values().filter { it != PresetCategories.MY_SAYINGS }
-                .mapIndexed { i, presetCategory ->
-                    var presetDto = presetDtos.firstOrNull { it.categoryId == presetCategory.id }
-                    if (presetDto == null) {
-                        presetDto = PresetCategoryDto(presetCategory.id, false, i)
-                        database.presetCategoryDao().insertPresetCategory(presetDto)
-                    }
-                    Category.PresetCategory(
-                        presetCategory.id,
-                        presetDto.sortOrder,
-                        presetDto.hidden,
-                        presetCategory.getNameId()
-                    )
-                }
+                .filter { presetCategory -> dbPresets.none { it.categoryId == presetCategory.id } }
+                .map { presetCategory -> populatePresetCategory(presetCategory.id) }
         }
     }
 
+    private suspend fun populatePresetCategory(categoryId: String) {
+        val presetCategory = PresetCategories.values().firstOrNull { it.id == categoryId } ?: error(
+            "Unknown preset category id: $categoryId"
+        )
+        val newPresetDto = PresetCategoryDto(categoryId, false, presetCategory.initialSortOrder)
+        database.presetCategoryDao().insertPresetCategory(newPresetDto)
+        Category.PresetCategory(
+            newPresetDto.categoryId,
+            newPresetDto.sortOrder,
+            newPresetDto.hidden,
+            presetCategory.getNameId()
+        )
```

Comment:
Doesn't look like this `Category.PresetCategory` is used

Reply:
You're absolutely right

Reply:
I swear this isn't me making an excuse, but I've definitely been leveraging copilot more lately, and I need to be better about reviewing what it spits out. I've been leaning on "it passes the tests, so it's good", but that can be confusing if it's adding in nonsense code like this.

Reply:
No you just need a second copilot to auto-review for nonsense code 😎